e2e/vaultsecrets/vaultsecrets.go

```diff
@@ -0,0 +1,288 @@
+	f.NoError(err)
+
+	renderedCert, err := waitForAllocSecret(allocID, "task", "/secrets/certificate.crt",
+		func(out string) bool {
+			return strings.Contains(out, "BEGIN CERTIFICATE")
+		}, wc)
+	f.NoError(err)
+
+	_, err = waitForAllocSecret(allocID, "task", "/secrets/access.key",
+		func(out string) bool {
+			return strings.Contains(out, secretValue)
+		}, wc)
+	f.NoError(err)
+
+	var re = regexp.MustCompile(`VAULT_TOKEN=(.*)`)
+
+	// check vault token was written and save it for later comparison
+	out, err = e2e.AllocExec(allocID, "task", "env", ns, nil)
+	f.NoError(err)
+	match := re.FindStringSubmatch(out)
+	f.NotNil(match, fmt.Errorf("could not find VAULT_TOKEN, got:%v\n", out))
+	taskToken := match[1]
+
+	// Update secret
+	out, err = e2e.Command("vault", "kv", "put",
+		fmt.Sprintf("%s/myapp", tc.secretsPath), "key=UPDATED")
+	f.NoError(err, out)
+
+	elapsed := time.Now().Sub(ttlStart)
+	time.Sleep((time.Second * 60) - elapsed)
+
+	// tokens will not be updated
+	out, err = e2e.AllocExec(allocID, "task", "env", ns, nil)
+	f.NoError(err)
+	match = re.FindStringSubmatch(out)
+	f.NotNil(match, fmt.Errorf("could not find VAULT_TOKEN, got:%v\n", out))
+	f.Equal(taskToken, match[1])
+
+	// cert will be renewed
+	_, err = waitForAllocSecret(allocID, "task", "/secrets/certificate.crt",
+		func(out string) bool {
+			return strings.Contains(out, "BEGIN CERTIFICATE") &&
+				out != renderedCert
+		}, wc)
+	f.NoError(err)
+
+	// secret will *not* be renewed because it doesn't have a lease to expire
+	_, err = waitForAllocSecret(allocID, "task", "/secrets/access.key",
+		func(out string) bool {
+			return strings.Contains(out, secretValue)
+		}, wc)
+	f.NoError(err)
+
+}
+
+// We need to namespace the keys in the policy, so read it in and replace the
+// values of the policy names
+func writePolicy(policyID, policyPath, testID string) (string, error) {
+	raw, err := ioutil.ReadFile(policyPath)
+	if err != nil {
+		return "", err
+	}
+	policyDoc := string(raw)
+	policyDoc = strings.ReplaceAll(policyDoc, "TESTID", testID)
+
+	cmd := exec.Command("vault", "policy", "write", policyID, "-")
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return "", err
+	}
+
+	go func() {
+		defer stdin.Close()
+		io.WriteString(stdin, policyDoc)
+	}()
+
+	out, err := cmd.CombinedOutput()
+	return string(out), err
+}
+
+// We need to namespace the vault paths in the job, so parse it
+// and replace the values of the template and vault fields
+func runJob(jobID, testID string, index int) error {
+
+	raw, err := ioutil.ReadFile("./vaultsecrets/input/secrets.nomad")
+	if err != nil {
+		return err
+	}
+	jobspec := string(raw)
+	jobspec = strings.ReplaceAll(jobspec, "TESTID", testID)
+	jobspec = strings.ReplaceAll(jobspec, "DEPLOYNUMBER", string(index))
+
+	return e2e.RegisterFromJobspec(jobID, jobspec)
+}
+
+// waitForAllocSecret is similar to e2e.WaitForAllocFile but uses `alloc exec`
+// to be able to read the secrets dir, which is not available to `alloc fs`
```

Comment:
isn't the secrets dir available from `$TASK/secrets`? 

```bash
$ nomad alloc fs 9cf40 redis/secrets
Mode        Size  Modified Time         Name
-rw-r--r--  23 B  2020-10-13T20:41:17Z  access.key
```

Reply:
Hm, that's interesting. I was going from our docs https://www.nomadproject.io/docs/runtime/environment#secrets:

> `secrets/`: This directory is private to each task, not accessible via the nomad alloc fs command or filesystem APIs and where possible backed by an in-memory filesystem. It can be used to store secret data that should not be visible outside the task.

Reply:
whoops. don't tell anyone.

Reply:
If that's just a stale/poor doc, I'll fix the docs and swap this out for `e2e.WaitForAllocFile` to save some code to maintain.

Reply:
I've verified that's not a behavior regression, but some confusing and/or aspirational documentation:

```
$ ./pkg/nomad-0.5.0 fs d87 redis/secrets
Mode        Size  Modified Time          Name
-rw-r--r--  0 B   10/14/20 12:29:34 UTC  wat.txt
```

Reply:
Oh, this is very interesting. I swapped out that function and I get...

```
=== RUN   TestE2E/VaultSecrets/*vaultsecrets.VaultSecretsTest/TestVaultSecrets
    vaultsecrets.go:208:
                Error Trace:    vaultsecrets.go:208
                Error:          Received unexpected error:
                                test for file content failed: got "Reading secret file prohibited: task/secrets/certificate.crt"
                                error: <nil>
                Test:           TestE2E/VaultSecrets/*vaultsecrets.VaultSecretsTest/TestVaultSecrets
=== RUN   TestE2E/Volumes
```

And that's because:

```
$ nomad alloc fs 0b6 task/secrets
Mode        Size     Modified Time         Name
-rw-r--r--  0 B      2020-10-14T12:37:17Z  .nomad-mount
-rw-r--r--  49 B     2020-10-14T12:37:17Z  access.key
-rw-r--r--  1.2 KiB  2020-10-14T12:38:13Z  certificate.crt
-rw-r--r--  26 B     2020-10-14T12:37:17Z  vault_token

$ nomad alloc fs 0b6 task/secrets/certificate.crt
Reading secret file prohibited: task/secrets/certificate.crt
```

So you can list the directory but _not_ read the file contents. I'll follow-up in another issue on this one.

Reply:
ohhhh.... I didn't even try that